parquet/src/main/java/org/apache/iceberg/data/parquet/BaseParquetReaders.java

```diff
@@ -148,96 +167,81 @@ public ParquetValueReader<?> struct(
+    public Optional<ParquetValueReader<?>> visit(StringLogicalTypeAnnotation stringLogicalType) {
+      return Optional.of(ParquetValueReaders.strings(desc));
     }
 
     @Override
-    public Optional<ParquetValueReader<?>> visit(
-        LogicalTypeAnnotation.EnumLogicalTypeAnnotation enumLogicalType) {
-      return Optional.of(new ParquetValueReaders.StringReader(desc));
+    public Optional<ParquetValueReader<?>> visit(EnumLogicalTypeAnnotation enumLogicalType) {
+      return Optional.of(ParquetValueReaders.strings(desc));
     }
 
     @Override
     public Optional<ParquetValueReader<?>> visit(DecimalLogicalTypeAnnotation decimalLogicalType) {
-      switch (primitive.getPrimitiveTypeName()) {
-        case BINARY:
-        case FIXED_LEN_BYTE_ARRAY:
-          return Optional.of(
-              new ParquetValueReaders.BinaryAsDecimalReader(desc, decimalLogicalType.getScale()));
-        case INT64:
-          return Optional.of(
-              new ParquetValueReaders.LongAsDecimalReader(desc, decimalLogicalType.getScale()));
-        case INT32:
-          return Optional.of(
-              new ParquetValueReaders.IntegerAsDecimalReader(desc, decimalLogicalType.getScale()));
-        default:
-          throw new UnsupportedOperationException(
-              "Unsupported base type for decimal: " + primitive.getPrimitiveTypeName());
-      }
+      return Optional.of(ParquetValueReaders.bigDecimals(desc));
     }
 
     @Override
-    public Optional<ParquetValueReader<?>> visit(
-        LogicalTypeAnnotation.DateLogicalTypeAnnotation dateLogicalType) {
+    public Optional<ParquetValueReader<?>> visit(DateLogicalTypeAnnotation dateLogicalType) {
       return Optional.of(dateReader(desc));
     }
 
     @Override
-    public Optional<ParquetValueReader<?>> visit(
-        LogicalTypeAnnotation.TimeLogicalTypeAnnotation timeLogicalType) {
-      return Optional.of(timeReader(desc, timeLogicalType.getUnit()));
+    public Optional<ParquetValueReader<?>> visit(TimeLogicalTypeAnnotation timeLogicalType) {
+      return Optional.of(timeReader(desc));
     }
 
     @Override
     public Optional<ParquetValueReader<?>> visit(
-        LogicalTypeAnnotation.TimestampLogicalTypeAnnotation timestampLogicalType) {
+        TimestampLogicalTypeAnnotation timestampLogicalType) {
       return Optional.of(
-          timestampReader(
-              desc,
-              timestampLogicalType.getUnit(),
-              ((Types.TimestampType) expected).shouldAdjustToUTC()));
+          timestampReader(desc, ((Types.TimestampType) expected).shouldAdjustToUTC()));
     }
 
     @Override
-    public Optional<ParquetValueReader<?>> visit(
-        LogicalTypeAnnotation.IntLogicalTypeAnnotation intLogicalType) {
+    public Optional<ParquetValueReader<?>> visit(IntLogicalTypeAnnotation intLogicalType) {
       if (intLogicalType.getBitWidth() == 64) {
+        if (intLogicalType.isSigned()) {
+          // this will throw an UnsupportedOperationException
```

Comment:
nit: Just wondering why not replace it with preconditions check like newly added code below, with that we can have a valid error message that "64 bit int logical type must be unsigned" . 

Now it just says unsupported logical type INTEGER and it might confuse the user.

Reply:
Updated.